balancer-js/src/modules/sor/pool-data/onChainData3.ts

```diff
@@ -0,0 +1,228 @@
+  'function getTokenRates() view returns (uint256, uint256)',
+  'function getWrappedTokenRate() view returns (uint256)',
+  'function getAmplificationParameter() view returns (uint256 value, bool isUpdating, uint256 precision)',
+  'function getPausedState() view returns (bool)',
+  'function inRecoveryMode() view returns (bool)',
+  'function getRate() view returns (uint256)',
+  'function getScalingFactors() view returns (uint256[] memory)', // do we need this here?
+  'function getPoolTokens(bytes32) view returns (address[], uint256[])',
+];
+
+const getTotalSupplyFn = (poolType: string) => {
+  if (poolType.includes('Linear') || ['StablePhantom'].includes(poolType)) {
+    return 'getVirtualSupply';
+  } else if (poolType === 'ComposableStable') {
+    return 'getActualSupply';
+  } else {
+    return 'totalSupply';
+  }
+};
+
+const getSwapFeeFn = (poolType: string) => {
+  if (poolType === 'Element') {
+    return 'percentFee';
+  } else if (poolType === 'FX') {
+    return 'protocolPercentFee';
+  } else {
+    return 'getSwapFeePercentage';
+  }
+};
+
+interface OnchainData {
+  poolTokens: [string[], string[]];
+  totalShares: string;
+  swapFee: string;
+  isPaused?: boolean;
+  inRecoveryMode?: boolean;
+  rate?: string;
+  scalingFactors?: string[];
+  weights?: string[];
+  targets?: [string, string];
+  wrappedTokenRate?: string;
+  amp?: [string, boolean, string];
+  tokenRates?: [string, string];
+}
+
+type GenericPool = SubgraphPoolBase;
+//  Omit<SubgraphPoolBase | Pool, 'tokens'> & {
+//   tokens: (SubgraphToken | PoolToken)[];
+// };
+
+const defaultCalls = (
+  id: string,
+  address: string,
+  vaultAddress: string,
+  poolType: string,
+  multicaller: Multicaller3
+) => {
+  multicaller.call(`${id}.poolTokens`, vaultAddress, 'getPoolTokens', [id]);
+  multicaller.call(`${id}.totalShares`, address, getTotalSupplyFn(poolType));
+  multicaller.call(`${id}.swapFee`, address, getSwapFeeFn(poolType));
+  // multicaller.call(`${id}.isPaused`, address, 'getPausedState');
+  // multicaller.call(`${id}.inRecoveryMode`, address, 'inRecoveryMode');
+  // multicaller.call(`${id}.rate`, address, 'getRate');
+  // multicaller.call(`${id}.scalingFactors`, address, 'getScalingFactors');
+};
+
+const weightedCalls = (
+  id: string,
+  address: string,
+  multicaller: Multicaller3
+) => {
+  multicaller.call(`${id}.weights`, address, 'getNormalizedWeights');
+};
+
+const linearCalls = (
+  id: string,
+  address: string,
+  multicaller: Multicaller3
+) => {
+  multicaller.call(`${id}.targets`, address, 'getTargets');
+  multicaller.call(`${id}.wrappedTokenRate`, address, 'getWrappedTokenRate');
+};
+
+const stableCalls = (
+  id: string,
+  address: string,
+  multicaller: Multicaller3
+) => {
+  multicaller.call(`${id}.amp`, address, 'getAmplificationParameter');
+};
+
+const gyroECalls = (id: string, address: string, multicaller: Multicaller3) => {
+  multicaller.call(`${id}.tokenRates`, address, 'getTokenRates');
```

Comment:
This should on be when pool.poolTypeVersion === 2?

Reply:
When `getTokenRates` is missing the call is simply ignored. We could add the check, but then adds complexity to the types. I'd rather keep it simple for now since we have just a few pools that would use a version check. Something to reconsider when we start sending more useless calls.